File: backend/src/workspace/domain/commands.py
# Standard Library
import uuid
from typing import Annotated

# External Libraries
from pydantic import AfterValidator, BeforeValidator, Field

# VerdanTech Source
from src import settings
from src.common.adapters.utils.spec_manager import SpecManager, Specs
from src.common.domain import Command
from src.garden.domain.commands import GardenKey

from .specs import specs

# Load all banned workspace names
banned_fields = []
with open(settings.static_path("banned_fields.txt"), "r") as file:
    for line in file:
        field = line.strip()
        banned_fields.append(field.lower())

# ======================================
# Fields
# ======================================

WorkspaceName = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "workspace_name")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["workspace_name"]["field"],
        json_schema_extra={
            "min_length": specs.values["workspace_name"][Specs.MIN_LENGTH],
            "max_length": specs.values["workspace_name"][Specs.MAX_LENGTH],
            "pattern": specs.values["workspace_name"][Specs.PATTERN],
        },
    ),
]
WorkspaceDescription = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "workspace_description")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["workspace_description"]["field"],
        json_schema_extra={
            "max_length": specs.values["workspace_description"][Specs.MAX_LENGTH],
        },
    ),
]
PlantingAreaName = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(SpecManager.get_validation_method(specs, "planting_area_name")),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["planting_area_name"]["field"],
        json_schema_extra={
            "min_length": specs.values["planting_area_name"][Specs.MIN_LENGTH],
            "max_length": specs.values["planting_area_name"][Specs.MAX_LENGTH],
            "pattern": specs.values["planting_area_name"][Specs.PATTERN],
        },
    ),
]
PlantingAreaDescription = Annotated[
    str,
    BeforeValidator(lambda v: v.strip()),
    AfterValidator(
        SpecManager.get_validation_method(specs, "planting_area_description")
    ),
    # Note: Field used only for annotation, to allow custom error messages.
    Field(
        description=specs.descriptions["planting_area_description"]["field"],
        json_schema_extra={
            "max_length": specs.values["planting_area_description"][Specs.MAX_LENGTH],
        },
    ),
]

# ======================================
# Commands
# ======================================


class WorkspaceCreateCommand(Command):
    """
    Creates a new workspace.
    """

    garden_key: GardenKey
    name: WorkspaceName
    description: WorkspaceDescription | None = None


class WorkspaceUpdateCommand(Command):
    """
    Updates a workspace.
    """

    workspace_ref: uuid.UUID
    name: WorkspaceName | None = None
    description: WorkspaceDescription | None = None


class WorkspaceDeleteCommand(Command):
    """
    Deletes a workspace.
    """

    workspace_ref: uuid.UUID


class PlantingAreaCreateCommand(Command):
    """
    Creates a new planting area on a workspace
    """

    workspace_ref: uuid.UUID
    name: PlantingAreaName
    description: PlantingAreaDescription | None = None


class PlantingAreaUpdateCommand(Command):
    """
    Updates a planting area.
    """

    workspace_ref: uuid.UUID
    planting_area_id: uuid.UUID
    name: PlantingAreaName | None = None
    description: PlantingAreaDescription | None = None


class PlantingAreaDeleteCommand(Command):
    """
    Deletes a planting area.
    """

    workspace_ref: uuid.UUID
    planting_area_id: uuid.UUID

File: frontend/src/lib/data/workspace/commands.ts
import { z as zod } from 'zod';
import { useMutation } from '@sveltestack/svelte-query';
import type {
	WorkspaceCreateCommand,
	WorkspaceUpdateCommand,
	WorkspaceDeleteCommand,
	PlantingAreaCreateCommand,
	PlantingAreaUpdateCommand,
	PlantingAreaDeleteCommand
} from '$codegen/types';
import {
	workspaceCreateCommandOp,
	workspaceDeleteCommandOp,
	workspaceUpdateCommandOp,
	plantingAreaCreateCommandOp,
	plantingAreaUpdateCommandOp,
	plantingAreaDeleteCommandOp
} from '$codegen';
import { workspaceFieldSchemas } from './schemas';
import { gardenFieldSchemas } from '$lib/data/garden/schemas';

/** Creates a new workspace in a garden. */
export const workspaceCreate = {
	schema: zod.object({
		garden_key: gardenFieldSchemas.key,
		name: workspaceFieldSchemas.workspace_name,
		description: workspaceFieldSchemas.workspace_description.optional()
	}),
	mutation: () => {
		return useMutation(function (data: WorkspaceCreateCommand) {
			return workspaceCreateCommandOp(data);
		});
	}
};

/** Updates a workspace. */
export const workspaceUpdate = {
	schema: zod.object({
		workspace_ref: zod.string().uuid(),
		name: workspaceFieldSchemas.workspace_name.optional(),
		description: workspaceFieldSchemas.workspace_description.optional()
	}),
	mutation: () => {
		return useMutation(function (data: WorkspaceUpdateCommand) {
			return workspaceUpdateCommandOp(data);
		});
	}
};

/** Deletes a workspace. */
export const workspaceDelete = {
	schema: zod.object({
		workspace_ref: zod.string().uuid()
	}),
	mutation: () => {
		return useMutation(function (data: WorkspaceDeleteCommand) {
			return workspaceDeleteCommandOp(data);
		});
	}
};

/** Creates a new planting area in a workspace. */
export const plantingAreaCreate = {
	schema: zod.object({
		workspace_ref: zod.string().uuid(),
		name: workspaceFieldSchemas.planting_area_name,
		description: workspaceFieldSchemas.planting_area_description.optional()
	}),
	mutation: () => {
		return useMutation(function (data: PlantingAreaCreateCommand) {
			return plantingAreaCreateCommandOp(data);
		});
	}
};

/** Updates a planting area. */
export const plantingAreaUpdate = {
	schema: zod.object({
		workspace_ref: zod.string().uuid(),
		planting_area_id: zod.string().uuid(),
		name: workspaceFieldSchemas.planting_area_name.optional(),
		description: workspaceFieldSchemas.planting_area_description.optional()
	}),
	mutation: () => {
		return useMutation(function (data: PlantingAreaUpdateCommand) {
			return plantingAreaUpdateCommandOp(data);
		});
	}
};

/** Deletes a planting area. */
export const plantingAreaDelete = {
	schema: zod.object({
		workspace_ref: zod.string().uuid(),
		planting_area_id: zod.string().uuid()
	}),
	mutation: () => {
		return useMutation(function (data: PlantingAreaDeleteCommand) {
			return plantingAreaDeleteCommandOp(data);
		});
	}
};
